File: core/src/wtable.cpp
// wtable.cpp — 除法表 GBV_DIV_H.bin 加载，复刻 poc/algo/wtable.py
#include "singan2/wtable.h"

#include <cstdint>
#include <fstream>
#include <iterator>
#include <stdexcept>
#include <string>
#include <vector>

namespace singan2 {

WTable load_w_table(const std::string& path) {
    std::ifstream fp(path, std::ios::binary);
    if (!fp) throw std::runtime_error("无法打开 w_Table 文件: " + path);

    std::vector<uint8_t> raw((std::istreambuf_iterator<char>(fp)), std::istreambuf_iterator<char>());
    WTable wt;
    wt.table.assign(W_TABLE_SIZE, 0);

    // 文件按 UINT16 小端读入；不足 W_TABLE_SIZE 的部分保留 0(与 poc np.pad constant_values=0 一致)
    const size_t n = std::min(raw.size() / 2, static_cast<size_t>(W_TABLE_SIZE));
    for (size_t k = 0; k < n; k++) {
        const uint16_t v =
            static_cast<uint16_t>(raw[k * 2]) |
            static_cast<uint16_t>(static_cast<uint16_t>(raw[k * 2 + 1]) << 8);  // 小端
        wt.table[k] = v;
    }
    return wt;
}

WTable gen_w_table() {
    WTable wt;
    wt.table.assign(W_TABLE_SIZE, 0);  // table[0] = 0
    for (int i = 1; i < W_TABLE_SIZE; i++) {
        double v = 65536.0 / i;
        int iv = static_cast<int>(v);  // 截断(与 numpy astype(uint16) 行为一致)
        if (iv > 65535) iv = 65535;
        if (iv < 0) iv = 0;
        wt.table[i] = static_cast<uint16_t>(iv);
    }
    return wt;
}

}  // namespace singan2

File: poc/algo/wtable.py
# -*- coding: utf-8 -*-
"""
wtable.py — 除法表 GBV_DIV_H.bin 加载（复刻 WinMain.cpp w_Read）

w_Table[n] 约为 65536/n（16位定点），用于代替除法。
读取: 整个文件按 UINT16 小端读入。
"""
import os
import numpy as np
import logging
import traceback

W_TABLE_SIZE = 16384  # 32768 bytes / 2


def get_logger():
    return logging.getLogger("wtable")


def load_w_table(file_path):
    """加载 GBV_DIV_H.bin -> w_Table (np.uint16 数组)。

    参数:
        file_path: GBV_DIV_H.bin 路径
    返回:
        np.ndarray(16384,) uint16
    异常:
        IOError: 文件读取失败
    """
    try:
        with open(file_path, "rb") as fp:
            raw = fp.read()
        if len(raw) != W_TABLE_SIZE * 2:
            get_logger().warning("w_Table 文件大小 %d (期望 %d)，仍尝试读取", len(raw), W_TABLE_SIZE * 2)
        n = min(len(raw) // 2, W_TABLE_SIZE)
        arr = np.frombuffer(raw[:n * 2], dtype="<u2").astype(np.uint16)
        if n < W_TABLE_SIZE:
            arr = np.pad(arr, (0, W_TABLE_SIZE - n), mode="constant", constant_values=0)
        get_logger().info("w_Table 加载完成: %d 项", n)
        return arr
    except Exception as exc:
        get_logger().error("加载 GBV_DIV_H.bin 异常: %s\n%s", exc, traceback.format_exc())
        raise


def gen_w_table():
    """生成理论 w_Table（65536/n），用于无 bin 文件时调试。"""
    n = np.arange(1, W_TABLE_SIZE + 1, dtype=np.float64)
    vals = (65536.0 / n)
    arr = np.clip(vals, 0, 65535).astype(np.uint16)
    # 索引0 为 0（C++ memset 后 w_Table[0]=0）
    return np.concatenate(([0], arr[:W_TABLE_SIZE - 1]))
